mirrord-layer/src/pod_api.rs

```diff
@@ -72,62 +74,158 @@ impl RuntimeData {
+            &impersonated_container_name,
+        )
+        .await;
+        let jobs_api: Api<Job> = Api::namespaced(client.clone(), &agent_namespace);
+
+        create_job_pod_agent(config, agent_image, &pods_api, runtime_data, &jobs_api).await?
+    };
+    pods_api
+        .portforward(&pod_name, &[61337])
+        .await
+        .map_err(LayerError::KubeError)
+}
+
+fn get_agent_name() -> String {
+    let agent_name = format!(
         "mirrord-agent-{}",
         Alphanumeric
             .sample_string(&mut rand::thread_rng(), 10)
             .to_lowercase()
     );
+    agent_name
+}
 
-    let agent_image = agent_image.unwrap_or_else(|| {
-        concat!("ghcr.io/metalbear-co/mirrord:", env!("CARGO_PKG_VERSION")).to_string()
-    });
+async fn create_ephemeral_container_agent(
+    config: LayerConfig,
+    agent_image: String,
+    pods_api: &Api<Pod>,
+) -> Result<String, LayerError> {
+    warn!("Ephemeral Containers is an experimental feature
+              >> Refer https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/ for more info");
+
+    let mirrord_agent_name = get_agent_name();
+
+    let ephemeral_container: EphemeralContainer = serde_json::from_value(json!({
+        "name": mirrord_agent_name,
+        "image": agent_image,
+        "imagePullPolicy": config.image_pull_policy,
+        "target_container_name": config.impersonated_container_name,
+        "env": [{"name": "RUST_LOG", "value": config.agent_rust_log}],
+        "command": [
+            "./mirrord-agent",
+            "-t",
+            "30",
+        ],
+    }))?;
+    debug!("Requesting ephemeral_containers_subresource");
+
+    let mut ephemeral_containers_subresource = pods_api
+        .get_subresource("ephemeralcontainers", &config.impersonated_pod_name)
+        .await
+        .map_err(LayerError::KubeError)?;
+
+    let mut spec = ephemeral_containers_subresource
+        .spec
+        .as_mut()
+        .ok_or_else(|| LayerError::PodSpecNotFound(config.impersonated_pod_name.clone()))?;
+
+    spec.ephemeral_containers = match spec.ephemeral_containers.clone() {
+        Some(mut ephemeral_containers) => {
+            ephemeral_containers.push(ephemeral_container);
+            Some(ephemeral_containers)
+        }
+        None => Some(vec![ephemeral_container]),
+    };
+
+    pods_api
+        .replace_subresource(
+            "ephemeralcontainers",
+            &config.impersonated_pod_name,
+            &PostParams::default(),
+            to_vec(&ephemeral_containers_subresource).unwrap(),
+        )
+        .await
+        .map_err(LayerError::KubeError)?;
+
+    let params = ListParams::default()
+        .fields(&format!("metadata.name={}", "mirrord_agent"))
+        .timeout(10);
+
+    let mut stream = pods_api
+        .watch(&params, "0")
+        .await
+        .map_err(LayerError::KubeError)?
+        .boxed();
+
+    while let Some(status) = stream.try_next().await? {
```

Comment:
to re-iterate, if you won't have timeout here and we'll get events that aren't error or added, this will run forever.

Reply:
seems overkill